### PR TITLE
💣🐒 Reduce webpackers' config monkey-patching and use v5 default

### PR DESCRIPTION
**Pull request by josemigallas**
> ⚠️ **IMPORTANT NOTE:**
> The main problem about upgrading webpacker and webpack is that at some point we stopped using the default configuration. Now webpacker 6 diverges too much, specially because it now supports webpack 5 and a lot changes there.
>
> This PR removes some "tech-debt" related to our webpack configuration. There are more to come.

Ran `webpacker install` to "re-install" all default files and configs and then cleaned up everything so that we have a standard webpacker installation. This way the upgrade to v6 and later shakapacker will happen smoothly.

### CODE REVIEW

File: config/webpack/environment.js [hunk 1]
```diff
@@ -1,44 +1,33 @@
...
 
-// Add global webpack configs here
-
-environment.loaders.delete('css')
-environment.loaders.delete('moduleCss')
-environment.loaders.delete('sass')
-environment.loaders.delete('moduleSass')
+// TS and TSX transpilation is done with tsc via ts-loader, we don't use babel even though it's the
+// webpacker's default. The only use for babel is to transpile legacy module Stats, which could not
+// be migrated to Typescript due to potential extreme sanity loss.
```
Comment (akostadinov) [inline config/webpack/environment.js:6]: could you clarify sanity loss? Sanity of the code or sanity of the coder?

File: config/webpack/environment.js [hunk 2]
```diff
@@ -62,4 +51,9 @@ Object.assign(fileLoader.use[0].options, {
   postTransformPublicPath: (p) => `window.rails_asset_host + ${p}`
 });
 
+// Whatever this is, it seems we don't need it.
+environment.loaders.delete('nodeModules')
+
+console.dir(environment.loaders, { depth: null })
```
Comment (akostadinov) [inline config/webpack/environment.js:57]: what is this doing?

**Review by akostadinov — APPROVED** [review]
Some questions and nitpicks that you may or may not ignore, at your discretion. This is a huge improvement!

File: config/webpack/environment.js [hunk 3]
```diff
@@ -62,4 +51,9 @@ Object.assign(fileLoader.use[0].options, {
   postTransformPublicPath: (p) => `window.rails_asset_host + ${p}`
 });
 
+// Whatever this is, it seems we don't need it.
+environment.loaders.delete('nodeModules')
```
Comment (akostadinov) [inline config/webpack/environment.js:55]: Probably this gives a little explanation:
https://github.com/rails/webpacker/blob/54c3ca9245e9ee330f8ca63b447c202290f7b624/docs/v4-upgrade.md#excluding-node_modules-from-being-transpiled-by-babel-loader

File: config/webpack/environment.js (hunk 1 above)
Comment (josemigallas) [reply config/webpack/environment.js:6]: > **Sanity.**
> the state of having a healthy mind and not being mentally ill

The coder of course 😆

File: .babelrc [hunk 4]
```diff
@@ -1,3 +1,4 @@
+// Only use to transpile JSX from app/javscript/src/Stats
```
[the remark this answers is not in the document]
Comment (josemigallas) [reply .babelrc:1]: I've been tempted of removing this file several times, since it is not used apparently. Anyway, it's not used anymore by webpacker and it's gonna be removed sooner than later in favor of `babel.config.js`. I wouldn't sweat about it.

File: config/webpack/environment.js (hunk 2 above)
Comment (josemigallas) [reply config/webpack/environment.js:57]: This is for debugging, I left it there by accident 😅

File: config/webpack/environment.js (hunk 3 above)
Comment (josemigallas) [reply config/webpack/environment.js:55]: You know what, this in fact should not be part of the PR. I removed it at first to "see" what happens but this loader comes from webpacker so why should we delete it. I will restore this line.

File: .postcssrc.yml
Comment (akostadinov) [inline .postcssrc.yml:1]: we need this file?

**Review by akostadinov — APPROVED** [review]
another nit pick but looks good!

File: .postcssrc.yml
Comment (josemigallas) [reply .postcssrc.yml:1]: Yes, now that we actually include (or rather don't remove) `postcss-loader`, it will complain without a config file. It complains anyway that we aren't using any plugin 😅 but it's gonna be changed to `postcss.config.js` soon and I'll try to make the plugins work as it is indented by webpacker.

File: config/webpack/environment.js (hunk 3 above)
Comment (akostadinov) [reply config/webpack/environment.js]: If it isn't needed, then it can speed up compilation to be removed. So maybe keep that line. But as you decide.

File: bin/webpack-dev-server [hunk 5]
```diff
@@ -4,16 +4,15 @@ ENV["RAILS_ENV"] ||= ENV["RACK_ENV"] || "development"
 ENV["NODE_ENV"]  ||= "development"
 
 require "pathname"
-
-APP_ROOT = File.expand_path("..", __dir__)
-
-ENV["BUNDLE_GEMFILE"] ||= File.expand_path("Gemfile", APP_ROOT)
+ENV["BUNDLE_GEMFILE"] ||= File.expand_path("../../Gemfile", Pathname.new(__FILE__).realpath)
```
Comment (jlledom) [inline bin/webpack-dev-server:7]: Why this change? it's not exactly the same?

File: babel.config.js
Comment (jlledom) [inline babel.config.js:1]: Is this file automatically generated?

File: config/initializers/content_security_policy.rb [hunk 6]
```diff
@@ -23,3 +23,7 @@
 # For further information see the following documentation:
 # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only
 # Rails.application.config.content_security_policy_report_only = true
+
+Rails.application.config.content_security_policy do |policy|
+  policy.connect_src :self, :https, "http://localhost:3035", "ws://localhost:3035" if Rails.env.development?
+end
```
Comment (jlledom) [inline config/initializers/content_security_policy.rb:29]: What's this for?

File: config/webpacker.yml [hunk 7]
```diff
@@ -3,60 +3,96 @@
...
+    - .ico
+    - .svg
+    - .eot
+    - .otf
+    - .ttf
+    - .woff
+    - .woff2
+
   extensions:
-    - .coffee
```
Comment (jlledom) [inline config/webpacker.yml:17]: what about coffee? is it not packed by webpack?

File: config/webpacker.yml [hunk 8]
```diff
@@ -3,60 +3,96 @@
...
+
   # Compile test packs to a separate directory
   public_output_path: packs-test
 
 production:
   <<: *default
 
   # Production depends on precompilation of packs prior to booting for performance.
-  compile: false
+  compile: true
```
Comment (jlledom) [inline config/webpacker.yml:92]: Why change this? I think it was correct before, we precompile on the image build and deliver the assets inside the image.

File: config/initializers/content_security_policy.rb (hunk 6 above)
Comment (akostadinov) [reply config/initializers/content_security_policy.rb:29]: Probably this is for `webpack-dev-server`. But please make sure app works well also without it as right now.

File: app/javascript/src/Stats/buyer/stats_application.jsx [hunk 9]
```diff
@@ -1,3 +1,4 @@
+/** @jsxRuntime classic */
```
[the remark this answers is not in the document]
Comment (josemigallas) [reply app/javascript/src/Stats/buyer/stats_application.jsx:1]: It's a pragma for babel's `plugin-transform-react-jsx`. In summary It's required to transpile `Stats/` files.

File: bin/webpack-dev-server (hunk 5 above)
Comment (josemigallas) [reply bin/webpack-dev-server:7]: Yes, but this is auto generated by the webpacker installer.

File: babel.config.js
Comment (josemigallas) [reply babel.config.js:1]: Yes

File: app/javascript/src/Stats/buyer/stats_application.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (josemigallas) [reply app/javascript/src/Stats/buyer/stats_application.jsx:1]: More context about this line: https://github.com/system-ui/theme-ui/issues/1160#issuecomment-715530924

File: config/initializers/content_security_policy.rb (hunk 6 above)
Comment (josemigallas) [reply config/initializers/content_security_policy.rb:29]: it's not needed 👍🏻

**Review by jlledom — COMMENTED** [review]
Why not upgrading `@types/jest`? Does it do any harm?

**Comment by josemigallas** [comment]
> Why not upgrading `@types/jest`? Does it do any harm?

It's not directly related to webpacker and I'd like to upgrade the whole jest suite in another PR.

**Review by jlledom — DISMISSED** [review]

**Comment by josemigallas** [comment]
This is not looking good. CI won't compile because OOM, the bundle is bigger and styles are not injected causing some tests to fail.
I'm gonna hold this for now and try upgrade directly to shakapacker, which has direct support over their Slack channel and is supposed to fix most of this stuff.
